File: src/NRustLightning.Server/Services/RPCClientProvider.cs
using System.Collections.Generic;
using System.Net.Http;
using Microsoft.Extensions.Options;
using NBitcoin.RPC;
using NRustLightning.Infrastructure.Configuration;
using NRustLightning.Infrastructure.Interfaces;
using NRustLightning.Infrastructure.Networks;
using NRustLightning.Server.Configuration;
using NRustLightning.Server.Interfaces;

namespace NRustLightning.Server.Services
{
    public class RPCClientProvider : IRPCClientProvider
    {
        Dictionary<string, RPCClient> _ChainClients = new Dictionary<string, RPCClient>();

        public RPCClientProvider(IOptions<Config> config, IHttpClientFactory httpClientFactory)
        {
            foreach (var chainConfig in config.Value.ChainConfiguration)
            {
                var rpc = chainConfig?.Rpc;
                if (!(rpc is null))
                {
                    rpc.HttpClient = httpClientFactory.CreateClient(nameof(RPCClientProvider));
                    _ChainClients.Add(chainConfig!.CryptoCode, rpc);
                }
            }
        }

        public IEnumerable<RPCClient> GetAll() => _ChainClients.Values;

        public RPCClient? GetRpcClient(string cryptoCode)
        {
            _ChainClients.TryGetValue(cryptoCode, out var rpc);
            return rpc;
        }

        public RPCClient? GetRpcClient(NRustLightningNetwork n) => GetRpcClient(n.CryptoCode);
    }
}

File: src/NRustLightning.GUI/src/NRustLightning.GUI.Server/RPCClientProvider.fs
namespace NRustLightning.GUI.Server

open System.Collections.Concurrent
open System.Net.Http
open Microsoft.Extensions.Options
open NBitcoin
open NBitcoin.RPC
open NRustLightning.GUI.Client.Configuration
open NRustLightning.Infrastructure.Configuration.SubConfiguration
open NRustLightning.Infrastructure.Interfaces
open NRustLightning.Infrastructure.Networks

type RPCClientProvider(opts: IOptionsMonitor<WalletBiwaConfiguration>, httpC: IHttpClientFactory,
                       networkProvider: NRustLightningNetworkProvider) as this =
    let clients = ConcurrentDictionary<string, RPCClient>()
    do
        for n in networkProvider.GetAll() do
            clients.TryAdd(n.CryptoCode, opts.CurrentValue.GetRPCClient(n.NBitcoinNetwork, httpC.CreateClient())) |> ignore
        opts.OnChange(this.OnOptionChange) |> ignore
    
    member private this.OnOptionChange(newOpts) =
        for n in networkProvider.GetAll() do
            let rpc = newOpts.GetRPCClient(n.NBitcoinNetwork, httpC.CreateClient())
            clients.AddOrReplace(n.CryptoCode, rpc)
        
    member this.GetRPCClient(cryptoCode: string) =
        match clients.TryGetValue(cryptoCode) with
        | true, v -> Some (v)
        | false, _ -> None
        
    member this.GetRPCClient(n: NRustLightningNetwork) =
        this.GetRPCClient(n.CryptoCode)

